### PR TITLE
rust classic benches with Criterion for ecntt/msm/ntt

### PR DESCRIPTION
**Pull request by vhnatyk**
## Describe the changes

Rust idiomatic benches for EC NTT, NTT, MSM

## Linked Issues

Resolves #

### CODE REVIEW

File: wrappers/rust/icicle-core/src/msm/mod.rs [hunk 1]
```diff
@@ -317,3 +317,143 @@ macro_rules! impl_msm_tests {
...
+
+                        let mut msm_results = DeviceVec::<Projective<C>>::cuda_malloc(batch_size).unwrap();
+                        let mut points_d = DeviceVec::<Affine<C>>::cuda_malloc(full_size).unwrap();
+                        points_d
+                            .copy_from_host_async(HostSlice::from_slice(&points_cloned), &stream)
+                            .unwrap();
+
+                        cfg.precompute_factor = precompute_factor as i32;
+
+                        cfg.precompute_factor = precompute_factor as i32;
```
Comment (ChickenLover) [inline wrappers/rust/icicle-core/src/msm/mod.rs:427]: double set

File: wrappers/rust/icicle-core/src/msm/mod.rs [hunk 2]
```diff
@@ -317,3 +317,143 @@ macro_rules! impl_msm_tests {
...
+                .unwrap_or(MAX_LOG2);
+
+            let stream = CudaStream::create().unwrap();
+            let mut cfg = MSMConfig::default();
+            cfg.ctx
+                .stream = &stream;
+            cfg.is_async = true;
+            cfg.large_bucket_factor = 5;
+            cfg.c = 4;
+            let precompute_factor = 8;
```
Comment (ChickenLover) [inline wrappers/rust/icicle-core/src/msm/mod.rs:384]: I think this line is not needed anymore

File: wrappers/rust/icicle-curves/icicle-bls12-381/Cargo.toml
Comment (ChickenLover) [inline wrappers/rust/icicle-curves/icicle-bls12-381/Cargo.toml:1]: Shouldn't we change it for all the other curves' Cargo.tomls?

File: wrappers/rust/icicle-curves/icicle-bn254/benches/msm.rs
Comment (ChickenLover) [inline wrappers/rust/icicle-curves/icicle-bn254/benches/msm.rs:1]: why only bls12 and bn254? What about grumpkin, bw and babybear?

**Review by ChickenLover — COMMENTED** [review]
Good job with the PR @vhnatyk! We needed these benches for a long time. Left some comments that I think we should address before the merge though

File: wrappers/rust/icicle-core/src/msm/mod.rs (hunk 1 above)
Comment (vhnatyk) [reply wrappers/rust/icicle-core/src/msm/mod.rs:427]: fixed

File: wrappers/rust/icicle-core/src/msm/mod.rs [hunk 3]
```diff
@@ -317,3 +317,143 @@ macro_rules! impl_msm_tests {
...
+                        let batch_size = 1 << batch_size_log2;
+                        let full_size = batch_size * test_size;
+
+                        if full_size > 1 << max_log2 {
+                            continue;
+                        }
+
+                        let mut scalars = <C::ScalarField as FieldImpl>::Config::generate_random(full_size);
+                        let scalars = <C::ScalarField as FieldImpl>::Config::generate_random(full_size);
+                        // a version of batched msm without using `cfg.points_size`, requires copying bases
```
[the remark this answers is not in the document]
Comment (vhnatyk) [reply wrappers/rust/icicle-core/src/msm/mod.rs:411]: cloning anyway is not part of benching. I don't quite understand the comment as it's just copied from tests, maybe you can clarify, should it be equal to size of single msm in the batch?

File: wrappers/rust/icicle-curves/icicle-bn254/benches/ntt.rs [hunk 4]
```diff
@@ -0,0 +1,5 @@
+use icicle_bn254::curve::ScalarField;
+
+use icicle_core::impl_ntt_bench;
+
+impl_ntt_bench!("BN254", ScalarField);
```
[the remark this answers is not in the document]
Comment (vhnatyk) [reply wrappers/rust/icicle-curves/icicle-bn254/benches/ntt.rs]: fixed

File: wrappers/rust/icicle-curves/icicle-bn254/benches/msm.rs
Comment (vhnatyk) [reply wrappers/rust/icicle-curves/icicle-bn254/benches/msm.rs:1]: added

File: wrappers/rust/icicle-curves/icicle-bls12-381/Cargo.toml
Comment (vhnatyk) [reply wrappers/rust/icicle-curves/icicle-bls12-381/Cargo.toml:1]: done

File: wrappers/rust/icicle-core/src/msm/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChickenLover) [reply wrappers/rust/icicle-core/src/msm/mod.rs:411]: While it is not a part of benching, it still makes running benches longer. Why do we need to clone bases here?

[the remark this answers is not in the document]
Comment (vhnatyk) [reply wrappers/rust/icicle-core/src/msm/mod.rs:411]: ok, not cloning points now

**Review by ChickenLover — APPROVED** [review]

**Review by jeremyfelder — APPROVED** [review]

**Pull request merged**